[mlir][fold-memref-alias-ops] Add support for folding memref.expand_shape involving dynamic dims

meshtag opened the pull request:
`fold-memref-alias-ops` bails out in presence of dynamic shapes in `memref.expand_shape` op. Handle this case.

bondhugula commented:
Please add a commit summary - it's blank now.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +63,99 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
+    for (int64_t i = 0; i < groupSize; ++i) {
       sizes[i] = expandShapeOp.getResultType().getDimSize(groups[i]);
-    SmallVector<int64_t> suffixProduct = computeSuffixProduct(sizes);
+      if (resultType.isDynamicDim(groups[i]))
```
break out of the loop if you found the dynamic dim?

prathameshpml replied:
We are using the static shape value directly as well ([here](https://github.com/meshtag/llvm-project/blob/224262a2d921da44c484146d8f5209a80a3306a5/mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp#L113)) whenever we can find it. In the event that we don't find it, we use `memref.dim` op.

meshtag replied:
We are using the static shape value directly as well ([here](https://github.com/meshtag/llvm-project/blob/224262a2d921da44c484146d8f5209a80a3306a5/mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp#L113)) whenever we can find it. In the event that we don't find it, we use `memref.dim` op.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +63,99 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
-        AffineMap::get(/*numDims=*/groupSize,
-                       /*numSymbols=*/0, srcIndexExpr),
-        dynamicIndices);
+    if (hasDynamicDims) {
```
Why are these different? You should be able to handle the dynamic and static cases similarly? That is one of the advantages of using `OpFoldResult` and `composedFoldedAffine*` methods.

meshtag replied:
Both cases are calling `computeSuffixProduct` and `linearize` with different call signatures. Respective affine maps are different as well. Is there a way to do it in one go?

MaheshRavishankar reviewed (changes requested):
Thanks! This looks really useful... have just a few questions though.

meshtag commented:
Gentle ping @MaheshRavishankar , @ftynse

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +63,99 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
+  ShapedType resultType = expandShapeOp.getResultType();
+
+  // Traverse all reassociation groups to determine the appropriate indice
+  // corresponding to each one of them post op folding.
```
I think all the logic here would be much simplified with https://github.com/llvm/llvm-project/commit/e095d978ba476c9624b4e72776089ea7301fa657 (it got reverted due to a bot failure, but should land soon enough). Maybe worth waiting for that?

meshtag replied:
Do you think this change will break anything in that patch? (I don't see any changes in this pass in that patch)
Is that patch introducing better analysis utilities? (which can probably help us deal with dynamic and static cases in one go? - as you mentioned above)

I have no problem waiting for that change to get in if it somehow helps us better this patch.

ftynse commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +63,99 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
+      // reassociation group.
+      SmallVector<Value> sizesVal(groupSize);
+      for (int64_t i = 0; i < groupSize; ++i) {
+        if (sizes[i] <= 0)
```
Please use `ShapedType::isDynamic` instead of the magic check for negative values.

meshtag commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +63,99 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
+      // Create affine expression variables for symbols in the newly constructed
+      // affine map.
+      SmallVector<AffineExpr> symbols(groupSize);
+      bindSymbolsList(ctx, MutableArrayRef{symbols});
```
I got this without it 
```
error: no matching function for call to ‘bindSymbolsList(mlir::MLIRContext*&, llvm::SmallVector<mlir::AffineExpr>&)’
  122 |       bindSymbolsList(ctx, symbols);
```

ftynse replied:
It's a function template without deduction guides, `bindSymbolsList<AffineExpr>(ctx, symbols)` should work.

ftynse commented on mlir/include/mlir/Dialect/MemRef/Utils/MemRefUtils.h:
```diff
@@ -64,6 +64,34 @@ getLinearizedMemRefOffsetAndSize(OpBuilder &builder, Location loc, int srcBits,
+/// Assuming `sizes` is `[s0, .. sn]`, return the vector<Value>
+///   `[s1 * ... * sn, s2 * ... * sn, ..., sn, 1]`.
+///
+/// It is the caller's responsibility to provide valid values which are expected
+/// to be constants with index type or results of dimension extraction ops
+/// (for ex. memref.dim op).
```
This isn't actually a requirement of this function. It may be if it is called to insert IR inside an affine scope such as `affine.for` that imposes such requirements, but it would be the caller's responsibility to ensure that they construct valid IR in the end.

MaheshRavishankar reviewed (changes requested):
Since expand shape is now going to carry the shapes explicitly won't the logic here be much simpler?

meshtag commented:
> Since expand shape is now going to carry the shapes explicitly won't the logic here be much simpler?

I think that'll only change the way we get dynamic dimensions for analysis. Currently, we are getting them through the type (`memref.dim` over the resultant's type), but once that change gets in, we can get that information from the `output_shape` field. I don't expect it to change the logic however (we would still need to calculate the suffix product and then linearize those values to get the resultant map). 

If that PR has no blockers, I have no problem waiting for it to get in.

ftynse commented:
> If that https://github.com/llvm/llvm-project/pull/90040 has no blockers, I have no problem waiting for it to get in.

I have lifted my block on that PR.

meshtag commented:
@MaheshRavishankar, I've rebased this PR and resolved all merge conflicts. 

I tried using `getOutputShape` utility added in the `expand_shape` PR referenced above, it ended up generating the same IR which we have atm using `memref.dim` op. The current logic for this seems less convoluted to me as `getOutputShape` returns only dynamic shapes (not immediately obvious to me - maybe consider naming it as `getDynamicOutputShape`?), so we would have to keep an additional variable to track the number of dynamic dimensions accessed and then pass that each time we want to get a dynamic dimension. 

Let me know your thoughts on this.

MaheshRavishankar commented:
> @MaheshRavishankar, I've rebased this PR and resolved all merge conflicts.
> 
> I tried using `getOutputShape` utility added in the `expand_shape` PR referenced above, it ended up generating the same IR which we have atm using `memref.dim` op. The current logic for this seems less convoluted to me as `getOutputShape` returns only dynamic shapes (not immediately obvious to me - maybe consider naming it as `getDynamicOutputShape`?), so we would have to keep an additional variable to track the number of dynamic dimensions accessed and then pass that each time we want to get a dynamic dimension.
> 
> Let me know your thoughts on this.

I was thinking you can use https://github.com/llvm/llvm-project/blob/4b75fcf0a50f4be955b611e8e20d84d90ea133c8/mlir/lib/Dialect/Arith/Utils/Utils.cpp#L24 which should return a mixed static dynamic list like you are doing here. Isnt this what you are looking to do here as well?

meshtag commented:
> return a mixed static dynamic list like you are doing here. Isnt this what you are looking to do here as well?

We have separate control paths for static and dynamic values (because we need different maps and different utility functions to calculate suffix product and linearize variables). All of this related infra expects things to be either in `llvm::ArrayRef<int64_t>` or `llvm::ArrayRef<Value>`, so we require extra separation/casting (from `SmallVector<OpFoldResult>`, which is what the utility returns) if we use the above mentioned utility (another potentially overkill option might be to make analysis utilities deal with mixed values themselves - not all of whom are implemented in this PR). Additionally, the function also expects the input shape as `ArrayRef<OpFoldResult>`, which means an extra step/loop for dimension extraction and casting. 

I feel this does not compose well and would convolute the logic even more without reducing the code (since we anyway require separate paths for static and dynamic values). I feel it might've made sense to do that if we had a single control flow path. Unless it's a blocker to land this patch, I am hesitant on going through that path.

MaheshRavishankar commented:
IIUC what you are trying to do is effectively this

```
SmallVector<OpFoldResult> sizes = .... 
AffineExpr s0, s1;
bindSymbol(builder.getContext(), s0, s1);
OpFoldResult product = builder.getIndexAttr(1);
SmallVector<OpFoldResult> suffixProduct(sizes.size());
int index = sizes.size() - 1;
suffixProduct[index--] = product;
for (auto size : llvm::reverse(sizes.drop_front())) {
  product = affine::makeComposedFoldedAffineApply(builder, loc, s0 * s1, {product, size});
  suffixProduct[index--] = product;
}
return suffixProduct.
```

The mixed static dynamic cases will all be handled automatically? and the `sizes` is what is returned by the `inferOutputShape`. I dont see why the logic needs to be duplicated.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Transforms/FoldMemRefAliasOps.cpp:
```diff
@@ -63,39 +64,85 @@ resolveSourceIndicesExpandShape(Location loc, PatternRewriter &rewriter,
+    }
+  }
+
+  auto outputShape = inferExpandShapeOutputShape(
```
Nit: Better to do this "before" the for loop above to avoid creating operations if no transformation is being done.

meshtag replied:
The above `for` loop is used to populate `srcShape` which is required for the function call. Are you talking about a different `for` loop?

MaheshRavishankar replied:
Ah my bad. Please ignore.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp:
```diff
@@ -155,5 +156,31 @@ void eraseDeadAllocAndStores(RewriterBase &rewriter, Operation *parentOp) {
+computeSuffixProductIRBlockImpl(Location loc, OpBuilder &builder,
+                                ArrayRef<OpFoldResult> sizes,
+                                OpFoldResult unit) {
+  if (sizes.empty())
```
This could be folded in. Dont see a reason for the split.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/Utils/MemRefUtils.cpp:
```diff
@@ -155,5 +156,31 @@ void eraseDeadAllocAndStores(RewriterBase &rewriter, Operation *parentOp) {
+  AffineExpr s0, s1;
+  bindSymbols(builder.getContext(), s0, s1);
+
+  for (int64_t r = strides.size() - 2; r >= 0; --r) {
```
Nit: I find it easier to read in this form
```
for (int64_t r = strides.size() - 1 ; r > 0; --r) {

}
```

MaheshRavishankar reviewed (approved):
Much cleaner! Thanks!

meshtag commented:
Can someone help me land this commit. Thanks.

MaheshRavishankar commented:
> Can someone help me land this commit. Thanks.

I can help land, but can you check if the error is from this PR or is unrelated.

Commit:
[mlir][fold-memref-alias-ops] Add support for folding memref.expand_shape involving dynamic dims

fold-memref-alias-ops pass bails out in presence of dynamic shapes which leads to unwanted propagation of alias types during other transformations. This can percolate down further and can lead to errors which should not have been created in the first place.

meshtag commented:
@MaheshRavishankar, the builds are passing now (failure was unrelated to my change, simply updating the branch did the trick).
